### PR TITLE
Adjustable file locations

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following (please check boxes, eg [X]) _Failure to fill the template will close your PR_:**

***Please submit all pull requests against the `development` branch. Failure to do so will delay or deny your request***

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/FTL/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:** 

## 10

---

This PR increases the flexibility concerning the files FTL will use. It adds a number of additional config options, listed below with their default value:

```
LOGFILE=/var/log/pihole-FTL.log
PIDFILE=/var/run/pihole-FTL.pid
PORTFILE=/var/run/pihole-FTL.port
SOCKETFILE=/var/run/pihole/FTL.sock
WHITELISTFILE=/etc/pihole/whitelist.txt
BLACKLISTFILE=/etc/pihole/blacklist.txt
GRAVITYFILE=/etc/pihole/gravity.list
REGEXLISTFILE=/etc/pihole/regex.list
SETUPVARSFILE=/etc/pihole/setupVars.conf
AUDITLISTFILE=/etc/pihole/auditlog.list
```

Furthermore, we will now check in three different places for a FTL config file before we go ahead and use the default values (as listed above):

1. `/etc/pihole/pihole-FTL.conf`
2. `/var/snap/pihole/common/etc/pihole/pihole-FTL.conf`
3. a file `pihole-FTL.conf` in the current directory (`$PWD`)

**Although this PR is made to release/4.1, it should not be understood as a blocker for the release.**

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

### CODE REVIEW

File: FTL.h [hunk 1]
```diff
@@ -93,23 +93,21 @@ enum { BLOCKING_DISABLED, BLOCKING_ENABLED, BLOCKING_UNKNOWN };
 // Static structs
 typedef struct {
 	const char* conf;
-	const char* log;
-	const char* pid;
-	const char* port;
+	const char* conf2;
```
Comment (Mcat12) [inline FTL.h]: This should be given a more descriptive name, such as `snapConf`

File: config.c [hunk 2]
```diff
@@ -253,6 +325,35 @@ void read_FTLconf(void)
 		fclose(fp);
 }
 
+void getpath(FILE* fp, const char *option, const char *defaultloc, char **pointer)
+{
+	char *buffer = parse_FTLconf(fp, "PIDFILE");
+
+	errno = 0;
+	// Use sscanf() to obtain filename from config file parameter only if buffer != NULL
+	if(!(buffer != NULL && sscanf(buffer, "%127ms", pointer)))
```
Comment (Mcat12) [inline config.c]: This logic can be simplified (De Morgan's Law) to use a single or expression.

File: config.c [hunk 3]
```diff
@@ -253,6 +325,35 @@ void read_FTLconf(void)
 		fclose(fp);
 }
 
+void getpath(FILE* fp, const char *option, const char *defaultloc, char **pointer)
```
Comment (Mcat12) [inline config.c]: Could this function get some documentation (what it does, what the parameters mean, etc)?

File: config.c [hunk 4]
```diff
@@ -13,15 +13,60 @@
...
+		printf("Notice: Found no readable FTL config file");
+	}
+
+	// Read LOGFILE value if available
+	// defaults to: "/var/log/pihole-FTL.log"
+	buffer = parse_FTLconf(fp, "LOGFILE");
+
+	errno = 0;
+	// Use sscanf() to obtain filename from config file parameter only if buffer != NULL
+	if(!(buffer != NULL && sscanf(buffer, "%127ms", &FTLfiles.log)))
```
Comment (Mcat12) [inline config.c]: Logic can be simplified (see lower comment)

File: config.c [hunk 5]
```diff
@@ -13,15 +13,60 @@
...
+		printf("Notice: Found no readable FTL config file");
+	}
+
+	// Read LOGFILE value if available
+	// defaults to: "/var/log/pihole-FTL.log"
+	buffer = parse_FTLconf(fp, "LOGFILE");
+
+	errno = 0;
+	// Use sscanf() to obtain filename from config file parameter only if buffer != NULL
+	if(buffer == NULL || sscanf(buffer, "%127ms", &FTLfiles.log) == 0)
```
Comment (Mcat12) [inline config.c]: `sscanf` returns `EOF` on error, which is -1 (but the macro should be used instead). This may need to be used instead of zero.

Comment (DL6ER) [reply config.c]: If the option is, e.g. `LOGFILE=`, we read an empty string. From the `man` page it is not entirely clear if `sscanf()` would return `0` or `EOF` is returned if the string passed to the function is of length zero. You may be right. Shall we test for `EOF` or shall we be more general an do something like ` < 1` ?

Comment (Mcat12) [reply config.c]: Testing for `< 1` will probably do fine (we could also do `!= 1`).

**Pull request merged**